### PR TITLE
feat(quinn,quinn-udp): Avoid `socket2` and `std::net::UdpSocket` dependencies in `wasm32-unknown-unknown` target

### PR DESCRIPTION
**Pull request by matheus23**
## Changes

- Adds a `cfg_alias` `wasm_browser` for `all(target_family = "wasm", target_os = "unknown")` (matching `wasm*-*-unknown`)
- Makes `socket2` a target-dependent dependency
- Code depending on `std::net::UdpSocket` (e.g. `Runtime::wrap_udp_socket` or `Endpoint::rebind`) becomes gated as `not(wasm_browser)`
- adds `web_time` as target-dependent dependency in `quinn` to replace `std::time::{Instant, Duration}`, which panic in Wasm otherwise.

## Testing

- These changes were verified outside this PR with a custom `Runtime` implementation in the browser and nodejs.
  I'm not contributing this test (yet?) as discussed.
- The "test (wasm32-unknown-unknown)" github action job was adjusted with a check making sure there's no `(import "env" ...)` in the `quinn` `.wasm` file when built.
  This is a good, even if not quite sufficient effort in making sure the Wasm build doesn't break.

### CODE REVIEW

File: .github/workflows/rust.yml [hunk 1]
```diff
@@ -73,45 +73,35 @@ jobs:
...
-
-      - name: Install stable toolchain
-        uses: dtolnay/rust-toolchain@stable
-
-      - name: Add wasm target
-        run: rustup target add wasm32-unknown-unknown
-
-      - name: Install nodejs v20
-        uses: actions/setup-node@v4
+      - uses: actions/checkout@v4
```
Comment (djc) [inline .github/workflows/rust.yml:76]: What's the rationale for removing all the step names here? Would prefer to split these cosmetic/superficial changes into a separate commit.

File: quinn/src/runtime.rs [hunk 2]
```diff
@@ -6,18 +6,20 @@ use std::{
...
+use crate::Instant;
+
 /// Abstracts I/O and timer operations for runtime independence
 pub trait Runtime: Send + Sync + Debug + 'static {
     /// Construct a timer that will expire at `i`
     fn new_timer(&self, i: Instant) -> Pin<Box<dyn AsyncTimer>>;
     /// Drive `future` to completion in the background
     fn spawn(&self, future: Pin<Box<dyn Future<Output = ()> + Send>>);
     /// Convert `t` into the socket type used by this runtime
+    #[cfg(feature = "net")]
```
Comment (djc) [inline quinn/src/runtime.rs:22]: Hmm, I feel like trait methods that appear/disappear based on a Cargo feature aren't great. Are you still using the `Runtime` trait in your WASM setup?

File: quinn/src/tests.rs [hunk 3]
```diff
@@ -160,7 +158,7 @@ fn read_after_close() {
             .unwrap()
             .await
             .expect("connect");
-        tokio::time::sleep_until(Instant::now() + Duration::from_millis(100)).await;
+        tokio::time::sleep(Duration::from_millis(100)).await;
```
Comment (djc) [inline quinn/src/tests.rs:161]: This seems completely unrelated, ideally pull it out into a separate commit?

File: quinn-udp/src/lib.rs [hunk 4]
```diff
@@ -27,29 +27,35 @@
 #![warn(unreachable_pub)]
 #![warn(clippy::use_self)]
 
-#[cfg(unix)]
+use std::net::{IpAddr, Ipv6Addr, SocketAddr};
```
Comment (djc) [inline quinn-udp/src/lib.rs:30]: Does it still make sense to depend on quinn-udp in the WASM world?

**Comment by matheus23** [comment]
I realized this PR is missing some rationale, so I'm writing this on the top level for better visibility:

> Does it still make sense to depend on quinn-udp in the WASM world?

This was one of the first things I was thinking initially, too (and asked about in the crates discord).

The only definitions that `quinn` without the `net` feature needs from `quinn-udp` now are `BATCH_SIZE`, `RecvMeta`, `EcnCodepoint` and `Transmit`.

The fundamental problem IMO is that `AsyncUdpSocket` is tied to `quinn` and its `Runtime`, but then `AsyncUdpSocket` depends on e.g. `RecvMeta`. `quinn-udp` and `quinn` are thus tightly coupled at the moment.

The solution would be to separate out the type definitions of `RecvMeta`, etc. into another crate that both `quinn-udp` and `quinn` depend on. I didn't want to do this as that's quite an invasive refactor.
It is *not* possible to move `RecvMeta` etc. to `quinn`, because then `quinn-udp` would like to depend on `quinn` and the other way around.

> Are you still using the `Runtime` trait in your WASM setup?

Yes, there's no way of using `quinn` without providing a `Runtime` (e.g. the `Endpoint` keeps a reference to it).

The alternative would be integrating with `quinn-proto` only, which basically means reimplementing `quinn`s connection and endpoint drivers, timers, API, etc. That's practically not an option.

File: quinn/src/runtime.rs (hunk 2 above)
Comment (matheus23) [reply quinn/src/runtime.rs:22]: I've gone back and forth between having this gated on a feature and having this gated on the `wasm32-unknown-unknown` target.

Another option is to keep this method but then error out in all cases if invoked. This looks clean on the `quinn` side, but is error prone when e.g. some new `quinn` contribution suddenly makes use of the `wrap_udp_socket` fn outside of the `net` feature.

File: .github/workflows/rust.yml (hunk 1 above)
Comment (matheus23) [reply .github/workflows/rust.yml:76]: 1. I introduced the step names in the old PR but don't really think they're helping much.
2. All other jobs don't use step names. This is more consistent.

**Review by djc — APPROVED** [review]

**Review by djc — APPROVED** [review]
Okay, I think this all makes sense. Thanks!

File: quinn-udp/Cargo.toml [hunk 5]
```diff
@@ -14,23 +14,27 @@ workspace = ".."
...
 
 [features]
-default = ["tracing", "log"]
+default = ["tracing", "log", "net"]
 # Configure `tracing` to log events via `log` if no `tracing` subscriber exists.
 log = ["tracing/log"]
 direct-log = ["dep:log"]
 # Use private Apple APIs to send multiple packets in a single syscall.
 fast-apple-datapath = []
+net = ["dep:socket2"]
```
Comment (dignifiedquire) [inline quinn-udp/Cargo.toml:23]: maybe add a short description of what exactly the feature does?

**Comment by djc** [comment]
@Ralith ping?

File: quinn-udp/src/lib.rs
Comment (mxinden) [inline quinn-udp/src/lib.rs:1]: Instead of the many `cfg(feature = "net")`, wouldn't something along the following be simpler?

``` rust
#[cfg(feature = "net")]
use net::*;

#[cfg(feature = "net")]
mod net {
  // Everything that was previously feature flagged behind net.
}
```

**Review by mxinden — COMMENTED** [review]
I understand the rational above for importing `quinn-udp` in a WASM context.

That said, for what my opinion is worth here, I can't judge whether it is worth the complexity it introduces. Will defer to the actual Quinn maintainers.

File: quinn-udp/src/lib.rs
Comment (matheus23) [reply quinn-udp/src/lib.rs:1]: I've got a version where I've experimented with that.
It's a much more invasive change, I'm not sure if it's much better?
I can add it to this PR if other people agree: https://github.com/n0-computer/quinn/pull/23/commits/d06b71b9772bb828bf02b21796825a2ba6b7887b

Comment (mxinden) [reply quinn-udp/src/lib.rs:1]: Thanks for the commit. I didn't think of the fact that `unix.rs` and `windows.rs` would need to be under the `net/` module.

At this point I don't have a better idea, nor a preference for the options listed here.

**Comment by Ralith** [comment]
You mentioned that you also considered gating on target instead of a cargo feature. Can you elaborate on why you went with the feature instead? I think exposing a subset API on some targets is less hazardous than a cargo feature, particularly as this is technically a semver break affecting users who have `default-features = false` and use only a custom runtime.

File: quinn-udp/src/lib.rs
Comment (RealAstolfo) [reply quinn-udp/src/lib.rs:1]: couldnt windows and unix stuff be stuffed under a "system" "sys" or something like that.

**Comment by matheus23** [comment]
Main reason was that I noticed this was similar to e.g. tokio's `net` feature (something you disable when you build for Wasm similarly).
The `cfg`-gating also gets a little more complex, e.g:
```rs
#[cfg(all(not(all(target_family = "wasm", target_os = "unknown")), unix))]
use std::os::unix::io::AsFd;
```

But I see that you have cfg-aliases enabled for quinn-udp, so I'll try it with that.

**Comment by matheus23** [comment]
I've updated the PR to use a cfg alias `wasm_browser`, which stands for `all(target_family = "wasm", target_os = "unknown")`.
With this, a bunch of cfgs get simpler. E.g. `all(not(wasm_browser), unix)` simplifies to just `unix`, since `wasm_browser` and `unix` are exclusive.

File: quinn-udp/src/lib.rs
Comment (matheus23) [reply quinn-udp/src/lib.rs:1]: The newest version that avoids the feature in favor of a cfg alias `wasm_browser` makes lots of these cfgs simpler, as that cfg is exclusive with unix and windows cfgs.
Please take a look at the latest diff @mxinden I think you'll like it more.

**Comment by matheus23** [comment]
The CI failed with:

> `/usr/bin/ld: final link failed: No space left on device`

**Comment by gretchenfrage** [comment]
> The CI failed with:
> 
> > `/usr/bin/ld: final link failed: No space left on device`

[Indeed](https://github.com/quinn-rs/quinn/actions/runs/12596988919/job/35135265022?pr=2037)... Re-running it, it would be convenient if it's spurious but exciting if it's not.

**Comment by gretchenfrage** [comment]
@matheus23 I know you saw this on Discord, but could you add [this commit](https://github.com/quinn-rs/quinn/pull/2124/commits/bc58859de84088a49a67f7834ac836bccf583c5d) to the PR by whatever means you find most convenient to make the CI pass?

**Comment by gretchenfrage** [comment]
> The only definitions that `quinn` without the `net` feature needs from `quinn-udp` now are `BATCH_SIZE`, `RecvMeta`, `EcnCodepoint` and `Transmit`.
> 
> The fundamental problem IMO is that `AsyncUdpSocket` is tied to `quinn` and its `Runtime`, but then `AsyncUdpSocket` depends on e.g. `RecvMeta`. `quinn-udp` and `quinn` are thus tightly coupled at the moment.
> 
> The solution would be to separate out the type definitions of `RecvMeta`, etc. into another crate that both `quinn-udp` and `quinn` depend on. I didn't want to do this as that's quite an invasive refactor. It is _not_ possible to move `RecvMeta` etc. to `quinn`, because then `quinn-udp` would like to depend on `quinn` and the other way around.

To be honest, I'm not sure this feature gating approach is the best solution. I find it hard to follow, and I worry about our ability to remember why it is that way and what we need to do to keep respecting the needs of the WASM environment in the same way in the future.

If the problem is that a user might need to depend on these items that are used in both `quinn` and `quinn-udp`, while not having `quinn-udp` in the dependency tree (only having `quinn` in the dependency tree), I see 2 alternatives to feature gating as such:

- As you suggest, factor out some common dependency crate between `quinn` and `quinn-udp`. "Cleanest", but I understand the hesitation.
- Create duplicate versions of these items in `quinn` (e.g. essentially copy/paste `quinn_udp::RecvMeta` into somewhere in `quinn`), so that `AsyncUdpSocket` only references the versions defined in Quinn. Implement `From`/`Into` for conversion between them (little bit of boilerplate), and have `quinn::runtime::{tokio, async_io}` use those conversions.

@ralith @djc @mxinden thoughts on either of these suggestions?

**Comment by djc** [comment]
> To be honest, I'm not sure this feature gating approach is the best solution. I find it hard to follow, and I worry about our ability to remember why it is that way and what we need to do to keep respecting the needs of the WASM environment in the same way in the future.

IMO the feature gating approach is okay even if it's not the cleanest. Personally, I feel like both alternatives (splitting out yet another smaller crate, or duplicating `RecvMeta` and `Transmit`) are probably worse.

Would be good to get another round of @Ralith's review.

**Comment by djc** [comment]
@gretchenfrage what is necessary to get you to "Approve" this PR? Would be good to move forward.

**Review by gretchenfrage — COMMENTED** [review]
> @gretchenfrage what is necessary to get you to "Approve" this PR? Would be good to move forward.

I was hoping to abstain and let @Ralith give a second approval, if a second approval is necessary. I was skeptical of the approach but I'm not inclined to veto it--it should not be considered blocked by me. I probably could have communicated that better.

If my review is needed I can schedule another look soon.

**Review by Ralith — APPROVED** [review]

**Comment by Ralith** [comment]
LGTM, thanks! We can always revisit if someone comes up with a cleaner solution.

**Comment by gretchenfrage** [comment]
Thank you for working on this!

**Comment by matheus23** [comment]
Thank y'all for the reviews Railith, mxinden and djc.
Also specifically thank you to djc for keeping the ball up on this one :pray:
